### PR TITLE
feat: initial implementation

### PR DESCRIPTION
**Pull request by andrewrynhard**

### CODE REVIEW

**Comment by andrewrynhard** [comment]
@TimJones PTAL. This module should work in conjunction with https://github.com/autonomy/terraform-talos-security. I would like to improve this with conditionals in TF 1.12, so for now it is intentionally pretty limited in the things you can configure.

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,121 @@
+data "template_file" "master_0" {
```
Comment (TimJones) [inline main.tf]:
```suggestion
data "template_file" "master" {
```

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  count = "1"
```
Comment (TimJones) [inline main.tf]:
```suggestion
  count = "${length(var.master_ip_addresses)}"
```

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+  template = "${file("${path.module}/templates/init.yaml")}"
```
Comment (TimJones) [inline main.tf]:
```suggestion
  template = "${count.index == 0 ? file("${path.module}/templates/init.yaml") : file("${path.module}/templates/join-control-plane.yaml")}"
```

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+  vars {
+    # Kubernetes.
+
+    kubernetes_ca_crt      = "${base64encode(var.kubernetes_ca_crt)}"
+    kubernetes_ca_key      = "${base64encode(var.kubernetes_ca_key)}"
+    token                  = "${var.kubernetes_token}"
+    api_server_cert_sans   = "${join(", ", var.master_ip_addresses)}"
+    api_endpoint           = "${var.master_ip_addresses[0]}"
```
Comment (TimJones) [inline main.tf]:
```suggestion
    api_endpoint           = "${var.master_hostnames[count.index]}"
```

File: main.tf [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    control_plane_endpoint = "${var.master_ip_addresses[0]}"
```
Comment (TimJones) [inline main.tf]:
```suggestion
    control_plane_endpoint = "${var.master_hostnames[0]}"
```

File: main.tf [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    pod_subnet             = "${var.pod_subnet}"
+    service_subnet         = "${var.service_subnet}"
+
+    # Talos.
+
+    os_ca_crt                   = "${base64encode(var.talos_ca_crt)}"
+    os_ca_key                   = "${base64encode(var.talos_ca_key)}"
+    trustd_username             = "${var.trustd_username}"
+    trustd_password             = "${var.trustd_password}"
+    trustd_endpoints            = "[${var.master_ip_addresses[1]}]"
```
Comment (TimJones) [inline main.tf]:
```suggestion
    trustd_endpoints            = "[${element(var.master_hostnames, count.index + 1)}]"
```
I use the `element` function here as Terraform interpolation will automatically wrap when `count.index + 1` goes out-of-bounds.

File: main.tf [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    container_network_interface = "${var.container_network_interface_plugin}"
+  }
+}
+
+data "template_file" "master_1" {
```
Comment (TimJones) [inline main.tf]: Can drop this whole section.
```suggestion
```

File: main.tf [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    # Talos.
+
+    trustd_username             = "${var.trustd_username}"
+    trustd_password             = "${var.trustd_password}"
+    trustd_endpoints            = "[${var.master_ip_addresses[2]}]"
+    container_network_interface = "${var.container_network_interface_plugin}"
+  }
+}
+
+data "template_file" "master_2" {
```
Comment (TimJones) [inline main.tf]: Can drop this whole section.
```suggestion
```

File: outputs.tf [hunk 9]
```diff
@@ -0,0 +1,7 @@
+output "masters" {
+  value = [ "${data.template_file.master_0.rendered}", "${data.template_file.master_1.rendered}", "${data.template_file.master_2.rendered}" ]
```
Comment (TimJones) [inline outputs.tf]:
```suggestion
  value = [ "${data.template_file.master.*.rendered}" ]
```

File: variables.tf [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+variable "kubernetes_ca_crt" {
+  description = "The PEM encoded Kubernetes certificate"
+}
+
+variable "kubernetes_ca_key" {
+  description = "The PEM encoded Kubernetes private key"
+}
+
+variable "master_ip_addresses" {
```
Comment (TimJones) [inline variables.tf]:
```suggestion
variable "master_hostnames" {
  description = "A list of resolvable hostnames used for the master nodes"
```

File: main.tf [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    cluster_name           = "${var.cluster_name}"
+    dns_domain             = "${var.dns_domain}"
+    pod_subnet             = "${var.pod_subnet}"
+    service_subnet         = "${var.service_subnet}"
+
+    # Talos.
+
+    trustd_username             = "${var.trustd_username}"
+    trustd_password             = "${var.trustd_password}"
+    trustd_endpoints            = "[${join(", ", var.master_ip_addresses)}]"
```
Comment (TimJones) [inline main.tf]:
```suggestion
    trustd_endpoints            = "[${join(", ", var.master_hostnames)}]"
```

File: main.tf [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+  template = "${file("${path.module}/templates/join-worker.yaml")}"
+
+  vars {
+    # Kubernetes.
+
+    kubernetes_ca_crt      = "${base64encode(var.kubernetes_ca_crt)}"
+    kubernetes_ca_key      = "${base64encode(var.kubernetes_ca_key)}"
+    token                  = "${var.kubernetes_token}"
+    api_server_cert_sans   = "${join(", ", var.master_ip_addresses)}"
+    api_endpoint           = "${var.master_ip_addresses[0]}"
```
Comment (TimJones) [inline main.tf]:
```suggestion
    api_endpoint           = "${var.master_hostnames[0]}"
```

File: main.tf [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+    control_plane_endpoint = "${var.master_ip_addresses[0]}"
```
Comment (TimJones) [inline main.tf]:
```suggestion
    control_plane_endpoint = "${var.master_hostnames[0]}"
```

File: main.tf [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+    master_ip              = "${var.master_ip_addresses[0]}"
```
Comment (TimJones) [inline main.tf]:
```suggestion
    master_ip              = "${var.master_hostnames[0]}"
```

File: main.tf [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    pod_subnet             = "${var.pod_subnet}"
+    service_subnet         = "${var.service_subnet}"
+
+    # Talos.
+
+    os_ca_crt                   = "${base64encode(var.talos_ca_crt)}"
+    os_ca_key                   = "${base64encode(var.talos_ca_key)}"
+    trustd_username             = "${var.trustd_username}"
+    trustd_password             = "${var.trustd_password}"
+    trustd_endpoints            = "[${element(var.master_hostnames, count.index + 1)}]"
```
Comment (TimJones) [inline main.tf]:
```suggestion
    trustd_endpoints            = "[${count.index + 1 < length(var.master_hostnames) ? var.master_hostnames[count.index + 1] : ""}]"
```

**Comment by TimJones** [comment]
@andrewrynhard made some suggestions on how to support 1..n masters but I haven't tested it yet.

**Comment by andrewrynhard** [comment]
@TimJones I think this is gonna be a WIP until I can fix the current bug with trustd that I introduced in https://github.com/autonomy/talos/pull/406. Also, I'm thinking of breaking out the `masters` data template file into two distinct ones. In kubeadm there are three types of nodes. An initial node the bootstraps PKI and the initial control plane, a worker node that is joining _as a master_, and then the normal working joining the node. Three distinct configuration files are needed. It be more readable and self as documentation of we make this clear in our TF code. Thoughts?

**Comment by TimJones** [comment]
@andrewrynhard so `master` would be split into something like `init_master` with a `count = "1"` `template = "${file("${path.module}/templates/init.yaml")}` and `join_master` with `count = "${length(var.master_hostnames) - 1}"` `template = "${file("${path.module}/templates/join-control-plane.yaml")}`?
It'll be more work maintain over the complexity of selecting the template file based on `count.index`.
If you're OK with that then it's fine by me :)

**Comment by andrewrynhard** [comment]
Yeah there will be a little more complexity 🤔. Hmm. The tradeoff would be self-documented code that users code refer to if they want. I'm torn. If it only ends up being the changes you are proposing then I think its reasonable. Also, you have done 90% of this code now :D did you want me to close this out and you can resubmit so that I'm not blocking on anything?

**Comment by TimJones** [comment]
If having the increased verbosity of multiple `master` data templates means it's easier to on-board new contributors I think it's a good call, given that we're already vigilant in our reviews :wink:
Either way the output of the module to other modules is the same so :+1:

**Comment by andrewrynhard** [comment]
Sounds good.

File: variables.tf [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+  description = "The PEM encoded Kubernetes certificate"
+}
+
+variable "kubernetes_ca_key" {
+  description = "The PEM encoded Kubernetes private key"
+}
+
+variable "master_hostnames" {
+  description = "A list of resolvable hostnames used for the master nodes"
+  description = "A list of static IP address used for the master nodes"
```
Comment (TimJones) [inline variables.tf]:
```suggestion
```
Doubled-up `description`

**Pull request merged**